src/com/ms/silverking/cloud/dht/daemon/storage/convergence/ConvergenceController2.java

```diff
@@ -697,314 +706,320 @@ private void fetchKeys() {
+  }
+
+  public static void _incomingChecksumTree(UUIDBase uuid, ChecksumNode remoteTree, ConvergencePoint cp,
+      MessageGroupConnection connection) {
+    convergencePauseMutex.lock();
+    try {
+      if (convergencePaused) {
         try {
-            RawRetrievalResult  rawRetrievalResult;
-            
-            if (entry.getValue() == null) {
-                Log.warning("Couldn't find value for: ", entry);
-                return;
-            } else {
-                if (debug) {
-                    System.out.printf("Found %s\n", entry);
-                }
-            }
-            rawRetrievalResult = new RawRetrievalResult(RetrievalType.VALUE_AND_META_DATA);
-            rawRetrievalResult.setStoredValue(entry.getValue(), true, false, null);
-            StorageValueAndParameters   valueAndParameters;
-            
-            ByteBuffer  rawValueBuffer;
-            ByteBuffer  valueBuffer;
-            
-            //valueBuffer = (ByteBuffer)entry.getValue().duplicate().limit(rawRetrievalResult.getStoredLength());
-            rawValueBuffer = entry.getValue();
-            if (debug && true) {
-                System.out.printf("key %s buf %s storedLength %d uncompressedLength %d compressedLength %d\n", 
-                    entry, rawValueBuffer, rawRetrievalResult.getStoredLength(),  
-                    rawRetrievalResult.getUncompressedLength(), MetaDataUtil.getCompressedLength(rawValueBuffer, 0));
-                System.out.printf("rawValueBuffer %s\n", StringUtil.byteBufferToHexString(rawValueBuffer));
-            }
-            
-            valueBuffer = (ByteBuffer)rawValueBuffer.duplicate().position(
-                    rawValueBuffer.position() + MetaDataUtil.getDataOffset(rawValueBuffer, 0));
-            
-            // FUTURE - consider making the nsstore allow a put that just accepts the buffer as is
-            // to improve performance
-            
-            valueAndParameters = new StorageValueAndParameters(entry, valueBuffer, 
-                    rawRetrievalResult.getVersion(), 
-                    rawRetrievalResult.getUncompressedLength(), 
-                    MetaDataUtil.getCompressedLength(rawValueBuffer, 0), 
-                    rawRetrievalResult.getCCSS(), rawRetrievalResult.getChecksum(), 
-                    rawRetrievalResult.getCreator().getBytes(), rawRetrievalResult.getCreationTimeRaw(),
-                    PutOptions.noVersionRequired, PutOptions.noLock);
-                    nsStore.put(ImmutableList.of(valueAndParameters), emptyUserData, this);
-            
-            // FUTURE - preserve user data
-        } catch (CorruptValueException cve) {
-            Log.logErrorWarning(cve);
-            // FUTURE - let this replica know that it's bad
+          Log.warning("Storing paused checksum tree ", uuid);
+          pausedChecksumTrees.put(new PausedChecksumTree(uuid, remoteTree, cp, connection));
+        } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)